docs/assumptions.rst

```diff
@@ -1,63 +1,180 @@
 ================
-Assumptions
+Assumptions and Semantics
 ================
 
 In this section, we provide assumptions we make about the behavior of
-applications that use UnifyFS, and about how UnifyFS currently functions.
+applications that use UnifyFS and about the file system semantics of UnifyFS.
 
 ---------------------------
 System Requirements
 ---------------------------
 
-UnifyFS uses node-local storage devices, e.g., RAM and SSD, for storing the
-application data. Therefore, a system should support the following requirements
-to run UnifyFS.
+The system requirements to run UnifyFS are:
 
-    - A compute node is equipped with a local storage device that UnifyFS can
+    - Compute nodes must be equipped with local storage device(s) that UnifyFS can
       use for storing file data, e.g., SSD or RAM.
 
-    - An ability for UnifyFS to launch user-level daemon processes on compute
-      nodes, which run concurrently with user application processes
+    - The system must support the ability for UnifyFS user-level server processes
+      to run concurrently with user application processes on compute nodes.
 
 ---------------------------
 Application Behavior
 ---------------------------
 
-UnifyFS is specifically designed to support globally synchronous checkpointing
-workloads. In such a workload, the expected application behavior is as follows.
+UnifyFS is specifically designed to support the bulk synchronous I/O pattern
+that is typical in HPC applications, e.g., checkpoint/restart or output dumps.
+In bulk synchronous I/O, I/O operations occur in separate write and read phases,
+and files are not read and written simultaneously.
+For example, files are written during checkpointing (a write phase)
+and read during recovery/restart (a read phase).
+Additionally, parallel writes and reads to shared files occur systematically,
+where processes access computable, regular offsets of files, e.g., in strided or
+segmented access patterns, in contrast to random, interleaved, small writes and reads.
```

Comment:
maybe also call out "overlapped writes" in the list of things not to do in bulk synchronous mode?

Reply:
Just for the team... I wouldn't change the text here, but technically we should be able to handle random and interleaved writes just fine.

We could also handle small writes efficiently.  Chunk space utilization and our extent tree can run into problems.  For extreme cases, we have it on our roadmap that we could also support a stripe-based storage model like Lustre rather than using a log-based I/O model.  The stripe-based method eliminates the need for extents and it can fully pack the storage, though it means we need to transfer data bytes on sync operations rather than extent info.  The write data could still be buffered locally until the sync/laminate, though.

Small local reads should also be decent in our current model, but lots of small remote reads will be bad.

Reply:
On when to switch to a striped-based method, there are a few interesting options:
1) allow user to set that as a config
2) use log-I/O during write, reformat to a stripe-based layout at laminate
3) track whether the extents for a file consume more space than the data as the file is being written, and switch layouts on the fly

Reply:
I made some updates to introduce the concept of overlapping writes and to drive home that the result in that case is undefined.